backend/src/v2/driver/driver.go

```diff
@@ -1188,43 +1214,235 @@ func resolveInputs(ctx context.Context, dag *metadata.DAG, iterationIndex *int,
+			glog.V(4).Infof(
+				"Overriding currentTask, %v, output with currentTask's producer_subtask, %v, output.",
+				currentTask.TaskName(),
+				subTaskName,
+			)
+
+			// Reassign sub-task before running through the loop again.
+			currentTask = tasks[subTaskName]
+		} else {
+			cfg.inputs.ParameterValues[cfg.name] = outputParametersCustomProperty[taskOutput.GetOutputParameterKey()]
+			// Exit the loop.
+			currentSubTaskMaybeDAG = false
+		}
+	}
+
+	return nil
+}
+
+// resolveUpstreamArtifactsConfig is just a config struct used to store the
+// input parameters of the resolveUpstreamArtifacts function.
+type resolveUpstreamArtifactsConfig struct {
+	ctx           context.Context
+	artifactSpec  *pipelinespec.TaskInputsSpec_InputArtifactSpec
+	dag           *metadata.DAG
+	pipeline      *metadata.Pipeline
+	mlmd          *metadata.Client
+	inputs        *pipelinespec.ExecutorInput_Inputs
+	name          string
+	artifactError func(error) error
+}
+
+// resolveUpstreamArtifacts resolves input artifacts that come from upstream
+// tasks. These tasks can be components/containers, which is relatively
+// straightforward, or DAGs, in which case, we need to traverse the graph until
+// we arrive at a component/container (since there can be n nested DAGs).
+func resolveUpstreamArtifacts(cfg resolveUpstreamArtifactsConfig) error {
+	taskOutput := cfg.artifactSpec.GetTaskOutputArtifact()
+	if taskOutput.GetProducerTask() == "" {
+		return cfg.artifactError(fmt.Errorf("producer task is empty"))
+	}
+	if taskOutput.GetOutputArtifactKey() == "" {
+		cfg.artifactError(fmt.Errorf("output artifact key is empty"))
+	}
+	tasks, err := getDAGTasks(cfg.ctx, cfg.dag, cfg.pipeline, cfg.mlmd, nil)
+	if err != nil {
+		cfg.artifactError(err)
+	}
+
+	producer, ok := tasks[taskOutput.GetProducerTask()]
+	if !ok {
+		cfg.artifactError(
+			fmt.Errorf("cannot find producer task %q", taskOutput.GetProducerTask()),
+		)
+	}
+	glog.V(4).Info("producer: ", producer)
+	currentTask := producer
+	var outputArtifactKey string
+	currentSubTaskMaybeDAG := true
+	// Continue looping until we reach a sub-task that is NOT a DAG.
+	for currentSubTaskMaybeDAG {
+		glog.V(4).Info("currentTask: ", currentTask.TaskName())
+		// If the current task is a DAG:
+		if *currentTask.GetExecution().Type == "system.DAGExecution" {
+			// Get the sub-task.
+			outputArtifactsCustomProperty := currentTask.GetExecution().GetCustomProperties()["output_artifacts"]
+			// Deserialize the output artifacts.
+			var outputArtifacts map[string]*pipelinespec.DagOutputsSpec_DagOutputArtifactSpec
+			err := json.Unmarshal([]byte(outputArtifactsCustomProperty.GetStringValue()), &outputArtifacts)
+			if err != nil {
+				return err
 			}
-			// TODO(Bobgy): cache results
-			outputs, err := mlmd.GetOutputArtifactsByExecutionId(ctx, producer.GetID())
+			glog.V(4).Infof("Deserialized outputArtifacts: %v", outputArtifacts)
+			artifactSelectors := outputArtifacts["Output"].GetArtifactSelectors()
```

Comment:
Key "Output" only works when a component has a single output. We should loop over the keys here.
```
			for k, value := range outputArtifacts {
				glog.V(4).Infof("k: %v", k)
				glog.V(4).Infof("outputArtifacts[k]: %v", value)
				artifactSelectors := value.GetArtifactSelectors()
				for _, v := range artifactSelectors {
					glog.V(4).Infof("v: %v", v)
					glog.V(4).Infof("v.ProducerSubtask: %v", v.ProducerSubtask)
					glog.V(4).Infof("v.OutputArtifactKey: %v", v.OutputArtifactKey)
					subTaskName = v.ProducerSubtask
					outputArtifactKey = v.OutputArtifactKey
				}
			}
```

Reply:
Note: This change breaks the lookup loop. I think we'll have to go to full recursion.